File: tests/mhd/mhd_extract.py
#! /usr/bin/env python3
'''
Extracts subvolumes of MHD output of PAMHD.

Copyright 2016 Ilja Honkonen
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice, this
  list of conditions and the following disclaimer in the documentation and/or
  other materials provided with the distribution.

* Neither the name of copyright holders nor the names of their contributors
  may be used to endorse or promote products derived from this software
  without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


If called as the main program...
'''

'''
Reads simulation data from infile_name and writes it into outfile_name while excluding cells whose center is outside of given volume
'''
def extract(
	infile_name,
	outfile_name,
	min_x = float('-inf'),
	max_x = float('+inf'),
	min_y = float('-inf'),
	max_y = float('+inf'),
	min_z = float('-inf'),
	max_z = float('+inf')
):
	from os.path import exists, isfile
	from numpy import dtype, fromfile, int32, uint64, zeros

	if not exists(infile_name):
		exit('Path ' + infile_name + " doesn't exist")
	if not isfile(infile_name):
		exit('Path ' + infile_name + ' is not a file')

	with open(infile_name, 'rb') as infile:
		with open(outfile_name, 'wb') as outfile:

			file_version = fromfile(infile, dtype = 'uint64', count = 1)
			if file_version[0] > 1:
				exit('File ' + infile_name + ' is of unsupported version: ' + str(file_version[0]))
			sim_params = fromfile(infile, dtype = '4double', count = 1)
			endianness = fromfile(infile, dtype = 'uint64', count = 1)
			ref_lvl_0_cells = fromfile(infile, dtype = '3uint64', count = 1)
			max_ref_lvl = fromfile(infile, dtype = 'intc', count = 1)
			if max_ref_lvl[0] > 0:
				exit('Refinement level > 0 not supported')
			neighborhood_length = fromfile(infile, dtype = 'uintc', count = 1)
			periodicity = fromfile(infile, dtype = '3uint8', count = 1)
			geometry_id = fromfile(infile, dtype = 'intc', count = 1)
			if geometry_id[0] != int32(1):
				exit('Unsupported geometry: ' + str(geometry_id[0]))
			grid_start = fromfile(infile, dtype = '3double', count = 1)
			lvl_0_cell_length = fromfile(infile, dtype = '3double', count = 1)
			total_cells = fromfile(infile, dtype = 'uint64', count = 1)
			cell_ids_data_offsets = fromfile(infile, dtype = '2uint64', count = total_cells[0])

			# check which cells are inside requested volume
			max_x_i, max_y_i, max_z_i = 0, 0, 0
			min_x_i, min_y_i, min_z_i = ref_lvl_0_cells[0][0], ref_lvl_0_cells[0][1], ref_lvl_0_cells[0][2]
			out_cells = []
			out_cell_data = []
			cell_data_t = 'double, 3double, double, 3double, 3double, intc, intc, double, 3double, 3double, 3double'
			for item in cell_ids_data_offsets:
				cell_id = uint64(item[0])

				cell_id -= 1
				cell_index = (
					uint64(cell_id % ref_lvl_0_cells[0][0]),
					uint64(cell_id / ref_lvl_0_cells[0][0] % ref_lvl_0_cells[0][1]),
					uint64(cell_id / (ref_lvl_0_cells[0][0] * ref_lvl_0_cells[0][1]))
				)

				cell_center = (
					grid_start[0][0] + lvl_0_cell_length[0][0] * (0.5 + cell_index[0]),
					grid_start[0][1] + lvl_0_cell_length[0][1] * (0.5 + cell_index[1]),
					grid_start[0][2] + lvl_0_cell_length[0][2] * (0.5 + cell_index[2])
				)
				cell_id += 1

				if cell_center[0] < min_x or cell_center[0] > max_x or cell_center[1] < min_y or cell_center[1] > max_y or cell_center[2] < min_z or cell_center[2] > max_z:
					continue

				min_x_i = min(min_x_i, cell_index[0])
				min_y_i = min(min_y_i, cell_index[1])
				min_z_i = min(min_z_i, cell_index[2])
				max_x_i = max(max_x_i, cell_index[0])
				max_y_i = max(max_y_i, cell_index[1])
				max_z_i = max(max_z_i, cell_index[2])

				out_cells.append(cell_id)

				infile.seek(item[1], 0)
				out_cell_data.append(
					fromfile(
						infile,
						dtype = cell_data_t,
						count = 1
					)
				)

			file_version.tofile(outfile)
			sim_params.tofile(outfile)
			endianness.tofile(outfile)

			# write grid parameters corresponding to filtered grid
			new_ref_lvl_0_cells = ref_lvl_0_cells.copy()
			new_ref_lvl_0_cells[0][0] = max_x_i - min_x_i + 1
			new_ref_lvl_0_cells[0][1] = max_y_i - min_y_i + 1
			new_ref_lvl_0_cells[0][2] = max_z_i - min_z_i + 1
			new_ref_lvl_0_cells.tofile(outfile)

			max_ref_lvl.tofile(outfile)
			neighborhood_length.tofile(outfile)
			periodicity.tofile(outfile)
			geometry_id.tofile(outfile)

			new_grid_start = grid_start.copy()
			new_grid_start[0][0] += lvl_0_cell_length[0][0] * min_x_i
			new_grid_start[0][1] += lvl_0_cell_length[0][1] * min_y_i
			new_grid_start[0][2] += lvl_0_cell_length[0][2] * min_z_i
			new_grid_start.tofile(outfile)

			lvl_0_cell_length.tofile(outfile)

			new_total_cells = total_cells.copy()
			new_total_cells[0] = len(out_cells)
			new_total_cells.tofile(outfile)

			cell_data_start = outfile.tell()
			out_cell_ids_offsets = zeros(len(out_cells), dtype = '2uint64')
			for i in range(len(out_cells)):

				old_id = out_cells[i] - 1
				old_index = (
					uint64(old_id % ref_lvl_0_cells[0][0]),
					uint64(old_id / ref_lvl_0_cells[0][0] % ref_lvl_0_cells[0][1]),
					uint64(old_id / (ref_lvl_0_cells[0][0] * ref_lvl_0_cells[0][1]))
				)
				new_index = (
					old_index[0] - min_x_i,
					old_index[1] - min_y_i,
					old_index[2] - min_z_i
				)
				new_id = uint64(
					new_index[2] * new_ref_lvl_0_cells[0][0] * new_ref_lvl_0_cells[0][1]
					+ new_index[1] * new_ref_lvl_0_cells[0][0]
					+ new_index[0]
				)
				new_id += 1

				out_cell_ids_offsets[i][0] = new_id
				out_cell_ids_offsets[i][1] = cell_data_start + out_cell_ids_offsets.nbytes + dtype(cell_data_t).itemsize * i

			out_cell_ids_offsets.tofile(outfile)
			for d in out_cell_data:
				d.tofile(outfile)


if __name__ == '__main__':
	from argparse import ArgumentParser, ArgumentDefaultsHelpFormatter
	from sys import argv

	parser = ArgumentParser(
		formatter_class = ArgumentDefaultsHelpFormatter
	)
	parser.add_argument('files', metavar = 'F', nargs = '*', help = 'Names of input and output files to process (in1 out1 in2 out2 ...)')
	parser.add_argument('--min-x', type = float, default = float('-inf'), help = 'Include in output file(s) only cells with center x coordinate larger than MIN_X')
	parser.add_argument('--max-x', type = float, default = float('+inf'), help = 'Include in output file(s) only cells with center x coordinate smaller than MAX_X')
	parser.add_argument('--min-y', type = float, default = float('-inf'), help = 'Include in output file(s) only cells with center y coordinate larger than MIN_Y')
	parser.add_argument('--max-y', type = float, default = float('+inf'), help = 'Include in output file(s) only cells with center y coordinate smaller than MAX_Y')
	parser.add_argument('--min-z', type = float, default = float('-inf'), help = 'Include in output file(s) only cells with center z coordinate larger than MIN_Z')
	parser.add_argument('--max-z', type = float, default = float('+inf'), help = 'Include in output file(s) only cells with center z coordinate smaller than MAX_Z')
	args = parser.parse_args()

	if len(args.files) % 2 > 0:
		exit('Even number of file arguments required (in1 out1 in2 out2 ...)')

	if args.min_x >= args.max_x:
		exit('Minimum x coordinate must be less than maximum')
	if args.min_y >= args.max_y:
		exit('Minimum y coordinate must be less than maximum')
	if args.min_z >= args.max_z:
		exit('Minimum z coordinate must be less than maximum')

	for i in range(0, len(args.files), 2):
		extract(
			args.files[i],
			args.files[i + 1],
			min_x = args.min_x,
			max_x = args.max_x,
			min_y = args.min_y,
			max_y = args.max_y,
			min_z = args.min_z,
			max_z = args.max_z
		)

File: tests/mhd/mhd_extract.cpp
/*
Extracts subvolumes of MHD output of PAMHD.

Copyright 2016 Ilja Honkonen
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice, this
  list of conditions and the following disclaimer in the documentation and/or
  other materials provided with the distribution.

* Neither the name of copyright holders nor the names of their contributors
  may be used to endorse or promote products derived from this software
  without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/


#include "array"
#include "cstdint"
#include "cstdio"
#include "iostream"
#include "string"
#include "vector"

#include "boost/program_options.hpp"


using namespace std;


/*
Writes given volume from file named infile_name to outfile_name.
*/
void extract(
	const string& infile_name,
	const string& outfile_name,
	const array<double, 3>& vol_min,
	const array<double, 3>& vol_max
) {
	FILE
		*infile = fopen(infile_name.c_str(), "rb"),
		*outfile = fopen(outfile_name.c_str(), "wb");

	if (infile == nullptr) {
		throw runtime_error("Couldn't open file " + infile_name);
	}
	if (outfile == nullptr) {
		throw runtime_error("Couldn't open file " + outfile_name);
	}

	size_t items_read;

	uint64_t file_version;
	items_read = fread(&file_version, sizeof(file_version), 1, infile);
	if (items_read != 1) { throw runtime_error("Couldn't read file version"); }

	array<double, 4> sim_params;
	items_read = fread(sim_params.data(), sizeof(double), sim_params.size(), infile);
	if (items_read != sim_params.size()) { throw runtime_error("Couldn't read simulation parameters"); }

	uint64_t endianness;
	items_read = fread(&endianness, sizeof(endianness), 1, infile);
	if (items_read != 1) { throw runtime_error("Couldn't read endianness"); }

	array<uint64_t, 3> ref_lvl_0_cells;
	items_read = fread(ref_lvl_0_cells.data(), sizeof(uint64_t), ref_lvl_0_cells.size(), infile);
	if (items_read != ref_lvl_0_cells.size()) { throw runtime_error("Couldn't read grid length in cells"); }

	int max_ref_lvl;
	items_read = fread(&max_ref_lvl, sizeof(max_ref_lvl), 1, infile);
	if (items_read != 1) { throw runtime_error("Couldn't read maximum refinement level"); }
	if (max_ref_lvl > 0) { throw runtime_error("Maximum refinement level > 0 not supported"); }

	unsigned int neigh_len;
	items_read = fread(&neigh_len, sizeof(neigh_len), 1, infile);
	if (items_read != 1) { throw runtime_error("Couldn't read neighborhood length"); }

	array<uint8_t, 3> periodicity;
	items_read = fread(periodicity.data(), sizeof(uint8_t), periodicity.size(), infile);
	if (items_read != periodicity.size()) { throw runtime_error("Couldn't read grid periodicity"); }

	int geom_id;
	items_read = fread(&geom_id, sizeof(geom_id), 1, infile);
	if (items_read != 1) { throw runtime_error("Couldn't read geometry id"); }

	array<double, 3> grid_start;
	items_read = fread(grid_start.data(), sizeof(double), grid_start.size(), infile);
	if (items_read != grid_start.size()) { throw runtime_error("Couldn't read grid starting coordinate"); }

	array<double, 3> lvl_0_cell_len;
	items_read = fread(lvl_0_cell_len.data(), sizeof(double), lvl_0_cell_len.size(), infile);
	if (items_read != lvl_0_cell_len.size()) { throw runtime_error("Couldn't read initial cell length"); }

	uint64_t total_cells;
	items_read = fread(&total_cells, sizeof(total_cells), 1, infile);
	if (items_read != 1) { throw runtime_error("Couldn't read total number of cells"); }

	// TODO: don't read in all at once if not needed
	vector<uint64_t> cell_ids_offsets(2 * total_cells);
	items_read = fread(cell_ids_offsets.data(), sizeof(uint64_t), cell_ids_offsets.size(), infile);
	if (items_read != cell_ids_offsets.size()) { throw runtime_error("Couldn't read cell ids and data offsets"); }

	array<uint64_t, 3>
		min_index{
			numeric_limits<uint64_t>::max(),
			numeric_limits<uint64_t>::max(),
			numeric_limits<uint64_t>::max()
		},
		max_index{
			numeric_limits<uint64_t>::min(),
			numeric_limits<uint64_t>::min(),
			numeric_limits<uint64_t>::min()
		};

	vector<uint64_t> out_cells;
	using cell_data_t = array<uint8_t, sizeof(double) * 21 + sizeof(int) * 2>;
	vector<cell_data_t> out_data;
	for (size_t i = 0; i < cell_ids_offsets.size(); i += 2) {
		auto cell_id = cell_ids_offsets[i];

		cell_id--;
		const array<uint64_t, 3> cell_index{
			cell_id % ref_lvl_0_cells[0],
			cell_id / ref_lvl_0_cells[0] % ref_lvl_0_cells[1],
			cell_id / (ref_lvl_0_cells[0] * ref_lvl_0_cells[1])
		};
		cell_id++;

		const array<double, 3> cell_center{
			grid_start[0] + lvl_0_cell_len[0] * (0.5 + cell_index[0]),
			grid_start[1] + lvl_0_cell_len[1] * (0.5 + cell_index[1]),
			grid_start[2] + lvl_0_cell_len[2] * (0.5 + cell_index[2])
		};

		if (
			cell_center[0] < vol_min[0]
			or cell_center[0] > vol_max[0]
			or cell_center[1] < vol_min[1]
			or cell_center[1] > vol_max[1]
			or cell_center[2] < vol_min[2]
			or cell_center[2] > vol_max[2]
		) {
			continue;
		}

		for (size_t dim = 0; dim < 3; dim++) {
			min_index[dim] = min(min_index[dim], cell_index[dim]);
			max_index[dim] = max(max_index[dim], cell_index[dim]);
		}

		out_cells.push_back(cell_id);

		const auto offset = cell_ids_offsets[i + 1];
		fseek(infile, offset, SEEK_SET);

		cell_data_t data;
		items_read = fread(data.data(), sizeof(cell_data_t::value_type), data.size(), infile);
		if (items_read != data.size()) {
			throw runtime_error(
				"Couldn't read data of cell " + to_string(cell_id)
				+ " at offset " + to_string(offset)
			);
		}
		out_data.push_back(data);
	}

	size_t items_written;

	items_written = fwrite(&file_version, sizeof(file_version), 1, outfile);
	if (items_written != 1) { throw runtime_error("Couldn't write file version"); }

	items_written = fwrite(sim_params.data(), sizeof(double), sim_params.size(), outfile);
	if (items_written != sim_params.size()) { throw runtime_error("Couldn't write simulation parameters"); }

	items_written = fwrite(&endianness, sizeof(endianness), 1, outfile);
	if (items_written != 1) { throw runtime_error("Couldn't write endianness"); }

	const decltype(ref_lvl_0_cells) new_lvl_0_cells{
		max_index[0] - min_index[0] + 1,
		max_index[1] - min_index[1] + 1,
		max_index[2] - min_index[2] + 1
	};
	items_written = fwrite(new_lvl_0_cells.data(), sizeof(uint64_t), new_lvl_0_cells.size(), outfile);
	if (items_written != new_lvl_0_cells.size()) { throw runtime_error("Couldn't write grid length in cells"); }

	items_written = fwrite(&max_ref_lvl, sizeof(max_ref_lvl), 1, outfile);
	if (items_written != 1) { throw runtime_error("Couldn't write maximum refinement level"); }
	if (max_ref_lvl > 0) { throw runtime_error("Maximum refinement level > 0 not supported"); }

	items_written = fwrite(&neigh_len, sizeof(neigh_len), 1, outfile);
	if (items_written != 1) { throw runtime_error("Couldn't write neighborhood length"); }

	items_written = fwrite(periodicity.data(), sizeof(uint8_t), periodicity.size(), outfile);
	if (items_written != periodicity.size()) { throw runtime_error("Couldn't write grid periodicity"); }

	items_written = fwrite(&geom_id, sizeof(geom_id), 1, outfile);
	if (items_written != 1) { throw runtime_error("Couldn't write geometry id"); }

	const decltype(grid_start) new_grid_start{
		grid_start[0] + lvl_0_cell_len[0] * min_index[0],
		grid_start[1] + lvl_0_cell_len[1] * min_index[1],
		grid_start[2] + lvl_0_cell_len[2] * min_index[2]
	};
	items_written = fwrite(new_grid_start.data(), sizeof(double), new_grid_start.size(), outfile);
	if (items_written != new_grid_start.size()) { throw runtime_error("Couldn't write grid starting coordinate"); }

	items_written = fwrite(lvl_0_cell_len.data(), sizeof(double), lvl_0_cell_len.size(), outfile);
	if (items_written != lvl_0_cell_len.size()) { throw runtime_error("Couldn't write initial cell length"); }

	uint64_t new_total_cells = out_cells.size();
	items_written = fwrite(&new_total_cells, sizeof(new_total_cells), 1, outfile);
	if (items_written != 1) { throw runtime_error("Couldn't write total number of cells"); }

	const auto cell_data_start = ftell(outfile);
	vector<uint64_t> new_cell_ids_offsets;
	for (size_t i = 0; i < out_cells.size(); i++) {
		const auto old = out_cells[i] - 1;

		const array<uint64_t, 3>
			old_index{
				old % ref_lvl_0_cells[0],
				old / ref_lvl_0_cells[0] % ref_lvl_0_cells[1],
				old / (ref_lvl_0_cells[0] * ref_lvl_0_cells[1])
			},
			new_index{
				old_index[0] - min_index[0],
				old_index[1] - min_index[1],
				old_index[2] - min_index[2]
			};
		const auto new_id
			= new_index[2] * new_lvl_0_cells[0] * new_lvl_0_cells[1]
			+ new_index[1] * new_lvl_0_cells[0]
			+ new_index[0] + 1;

		new_cell_ids_offsets.push_back(new_id);
		new_cell_ids_offsets.push_back(
			cell_data_start
			+ 2 * out_cells.size() * sizeof(uint64_t)
			+ i * sizeof(cell_data_t::value_type) * tuple_size<cell_data_t>::value
		);
	}

	items_written = fwrite(
		new_cell_ids_offsets.data(),
		sizeof(uint64_t),
		new_cell_ids_offsets.size(),
		outfile
	);
	if (items_written != new_cell_ids_offsets.size()) { throw runtime_error("Couldn't write cell ids and data offsets"); }

	items_written = fwrite(
		out_data.data(),
		tuple_size<cell_data_t>::value * sizeof(cell_data_t::value_type),
		out_data.size(),
		outfile
	);
	if (items_written != out_data.size()) { throw runtime_error("Couldn't write cell data"); }
}


int main(int argc, char* argv[])
{
	// program options
	vector<string> input_files;
	double
		min_x = -numeric_limits<double>::infinity(),
		max_x = +numeric_limits<double>::infinity(),
		min_y = -numeric_limits<double>::infinity(),
		max_y = +numeric_limits<double>::infinity(),
		min_z = -numeric_limits<double>::infinity(),
		max_z = +numeric_limits<double>::infinity();

	boost::program_options::options_description
		options("Usage: program_name [options], where options are");
	options.add_options()
		("help", "Print usage instructions")
		("input-file",
			boost::program_options::value<vector<string>>(&input_files),
			"Files to process (in1 out1 in2 out2 ..., the --input-file part can be omitted")
		("min-x",
			boost::program_options::value<double>(&min_x)
				->default_value(min_x),
			"Minimum x coordinate of volume to extract")
		("max-x",
			boost::program_options::value<double>(&max_x)
				->default_value(max_x),
			"Maximum x coordinate of volume to extract")
		("min-y",
			boost::program_options::value<double>(&min_y)
				->default_value(min_y),
			"Minimum y coordinate of volume to extract")
		("max-y",
			boost::program_options::value<double>(&max_y)
				->default_value(max_y),
			"Maximum y coordinate of volume to extract")
		("min-z",
			boost::program_options::value<double>(&min_z)
				->default_value(min_z),
			"Minimum z coordinate of volume to extract")
		("max-z",
			boost::program_options::value<double>(&max_z)
				->default_value(max_z),
			"Maximum z coordinate of volume to extract");

	boost::program_options::positional_options_description positional_options;
	positional_options.add("input-file", -1);

	boost::program_options::variables_map var_map;
	try {
		boost::program_options::store(
			boost::program_options::command_line_parser(argc, argv)
				.options(options)
				.positional(positional_options)
				.run(),
			var_map
		);
	} catch (exception& e) {
		cerr <<  __FILE__ << "(" << __LINE__ << "): "
			<< "Couldn't parse command line options: " << e.what()
			<< endl;
		abort();
	}
	boost::program_options::notify(var_map);

	if (var_map.count("help") > 0) {
		cout << options << endl;
		return EXIT_SUCCESS;
	}

	for (size_t i = 0; i < input_files.size(); i += 2) {
		extract(
			input_files[i],
			input_files[i + 1],
			{min_x, min_y, min_z},
			{max_x, max_y, max_z}
		);
	}

	return EXIT_SUCCESS;
}
